### ISSUE TITLE
build-djgpp.sh requires sudo privileges to build and install in /usr/local/

### ISSUE THREAD

**Issue by volkertb** [andrewwutw/build-djgpp#30]
Hi there,

First of all, awesome project. This script really saves people a lot of time.

The issue I'm reporting here is a minor one: by default, the script installs the cross-compiler and tools in /usr/local/djgpp/. However, on most systems, /usr/local/ is not writable for non-root users, so the script will have to be run with sudo privileges, unless an alternative DJGPP_PREFIX somewhere in the user's home folder is specified.

It might be helpful to clarify this in the README.md file.

Thanks and keep up the good work!

**Comment by stsp** [comment]
I also ran into this problem.
I think the compilation and installation
should be done by separate scripts,
with the second one being run with sudo.

**Comment by volkertb** [comment]
I agree with @stsp. That would be in line with how most Unix/Linux makefiles work, with the compilation being done with "make" and the installation being done with "sudo make install" (or just "make install" when a non-privileged installation folder is specified).

**Comment by stsp** [comment]
There seem to be a very extensive build
script rework pending: #25.
@jwt27 could you please hint us if your
patch set addresses also this?

**Comment by jwt27** [comment]
I hadn't really considered this to be an issue. There is already a check in place to test if you have write access to the installation directory:

https://github.com/andrewwutw/build-djgpp/blob/4180d9298bf2a78b9eed0be545a25931e2ee91e6/script/7.2.0#L134-L146

https://github.com/jwt27/build-gcc/blob/a8b3e194289c4f29dc820a30c0b5a4bd52de30d9/script/download.sh#L127-L139

I suppose this check could be moved up earlier in the script, before the download phase. Splitting up into separate build/install scripts makes things more complicated, since building gcc requires binutils and the libc headers to be installed first.

**Comment by stsp** [comment]
> libc headers to be installed first.

Ahh, indeed. :(
I wonder how can this be then solved properly.
fakechroot? Too complicated.
Maybe the build scripts should run sudo
themselves before doing the install things,
and the user will have to type the password
from time to time? Doh.

**Comment by jwt27** [comment]
> Maybe the build scripts should run sudo
> themselves before doing the install things,
> and the user will have to type the password
> from time to time? Doh.

That's an option. I'll look into it.

**Comment by stsp** [comment]
Thank you!
Please note that it will do so only if the
write perms are missing by default, so
at least no one from current users would
get sufficiently hurt by such change.

**Comment by jwt27** [comment]
See branch `sudo` on my fork: https://github.com/jwt27/build-gcc/tree/sudo

I am unable to test this myself, being on mingw64 (where sudo doesn't even exist)

**Comment by stsp** [comment]
First problem: bash (at least on ubuntu)
doesn't seem to like DOSish line endings.
Had to do `dos2unix` on scripts before they
could be started.

**Comment by stsp** [comment]
```
If you wish to change anything, press CTRL-C now. Otherwise, press any other key to continue.
[sudo] password for stas: 
Download source files...
Download http://www.delorie.com/pub/djgpp/current/v2gnu/bnu2311s.zip ...
```

I wonder why it asked the password before even
starting the download. Does this make any sense?

**Comment by jwt27** [comment]
> First problem: bash (at least on ubuntu)
> doesn't seem to like DOSish line endings.
> Had to do dos2unix on scripts before they
> could be started.

That's interesting.. I thought git converts line endings automatically? Will need to check my settings.

> I wonder why it asked the password before even
> starting the download. Does this make any sense?

That is to create the PREFIX directory. You're right, it doesn't make much sense to do that before downloading.

**Comment by stsp** [comment]
> That is to create the PREFIX directory.

Maybe, but now I restarted the script, and
it did the same thing again! (asked the sudo
passwd beforehand). Presumably the PREFIX
dir was already created on the second start,
so there is some bug.

**Comment by stsp** [comment]
```
chmod a+x install-sh
chmod a+x missing
/home/stas/src/build-gcc/script/build-binutils.sh: line 9: ../configure: Permission denied
```

You forgot to chmod also configure.

**Comment by stsp** [comment]
I wonder how to give the user a confidence
that he isn't working as root most of the time.
Maybe after you ended the sudo operation,
you should print what was done under root,
and say that after a kepress the root privs
will be dropped. Or overkill?

**Comment by jwt27** [comment]
> Maybe, but now I restarted the script, and it did the same thing again!
That makes sense since it runs `sudo mkdir -p ${PREFIX}`. I'll have to add a check to see if PREFIX exists instead of relying on `-p`.

> You forgot to chmod also configure.
Thanks, fixed. I don't notice these things on windows....

**Comment by stsp** [comment]
```
checking for shl_load... configure: error: Link tests are not allowed after GCC_NO_EXECUTABLES.
```

and build failed.
It would be good if you disable the parallel
build for now, otherwise its not clear where
was an error.

**Comment by jwt27** [comment]
> I think you left set -x somewhere already

That's `unpack-gcc.sh` from djcross, not mine. I don't know why it's enabled there.

> `main::scan_file() called too early ...`

Whoa, where did this happen?

> cp: cannot stat '/usr/local/cross/bin/stubify': No such file or directory

That's a bug. fixed.

> It would be good if you disable the parallel build for now, otherwise its not clear where was an error.

Set `MAKE_JOBS=1` before running the script.

**Comment by stsp** [comment]
> Whoa, where did this happen?

Here:
```
patching file /home/stas/src/build-gcc/build/djcross-gcc-8.2.0/gnu/gcc-8.20/include/libiberty.h
Using Plan A...
Hunk #1 succeeded at 108.
done
main::scan_file() called too early to check prototype at /home/stas/src/build-gcc/build/djcross-gcc-8.2.0/tmpinst/bin/aclocal line 644.
aclocal: couldn't open directory '/home/stas/src/build-gcc/build/djcross-gcc-8.2.0/tmpinst/share/aclocal': No such file or directory
main::scan_file() called too early to check prototype at /home/stas/src/build-gcc/build/djcross-gcc-8.2.0/tmpinst/bin/aclocal line 644.
aclocal: couldn't open directory '/home/stas/src/build-gcc/build/djcross-gcc-8.2.0/tmpinst/share/aclocal': No such file or directory
main::scan_file() called too early to check prototype at /home/stas/src/build-gcc/build/djcross-gcc-8.2.0/tmpinst/bin/aclocal line 644.
aclocal: couldn't open directory '/home/stas/src/build-gcc/build/djcross-gcc-8.2.0/tmpinst/share/aclocal': No such file or directory
main::scan_file() called too early to check prototype at /home/stas/src/build-gcc/build/djcross-gcc-8.2.0/tmpinst/bin/aclocal line 644.
aclocal: couldn't open directory '/home/stas/src/build-gcc/build/djcross-gcc-8.2.0/tmpinst/share/aclocal': No such file or directory
```

> That's a bug. fixed.

But nothing changed.
I guess I need to remove what was already
built, and start from the beginning... sigh. :)

**Comment by jwt27** [comment]
Okay that's weird. Looks like something went wrong building automake? I haven't seen that before.

>I guess I need to remove what was already built, and start from the beginning... sigh. :)

You only need to remove the gcc directory: `rm build/djcross-gcc-8.2.0/`

**Comment by stsp** [comment]
```
Copy long name executables to short name.
/home/stas/src/build-gcc/script/finalize.sh: line 13: /usr/local/cross/setenv-i586-pc-msdosdjgpp: Permission denied
/home/stas/src/build-gcc/script/finalize.sh: line 14: /usr/local/cross/setenv-i586-pc-msdosdjgpp: Permission denied
/home/stas/src/build-gcc/script/finalize.sh: line 15: /usr/local/cross/setenv-i586-pc-msdosdjgpp: Permission denied
/home/stas/src/build-gcc/script/finalize.sh: line 16: /usr/local/cross/setenv-i586-pc-msdosdjgpp: Permission denied
/home/stas/src/build-gcc/script/finalize.sh: line 18: /usr/local/cross/setenv-i586-pc-msdosdjgpp.bat: Permission denied
/home/stas/src/build-gcc/script/finalize.sh: line 19: /usr/local/cross/setenv-i586-pc-msdosdjgpp.bat: Permission denied
/home/stas/src/build-gcc/script/finalize.sh: line 20: /usr/local/cross/setenv-i586-pc-msdosdjgpp.bat: Permission denied
```

**Comment by jwt27** [comment]
Fixed it.

**Comment by jwt27** [comment]
Ah, yes. forgot about those. Should be fixed now.

**Comment by stsp** [comment]
But your scripts still contain the wrong
line-endings, so I always have to manually
fix them.

**Comment by jwt27** [comment]
> But your scripts still contain the wrong line-endings, so I always have to manually fix them.

I fixed that just now by running dos2unix manually. No idea why git didn't auto-convert them, I thought I had that set up right. Will have to check my other repos too.

**Comment by stsp** [comment]
> I fixed that just now by running dos2unix manually.

Where?
I don't have that fix in branch `sudo`.

**Comment by jwt27** [comment]
I merged it in the main branch (`generic`) together with the changes from `sudo`.

**Comment by stsp** [comment]
Looks functional!
Thanks again.
